elftools/dwarf/die.py

```diff
@@ -224,60 +225,64 @@ def _parse_DIE(self):
-                value = self._translate_attr_value(form, raw_value)
-            else:
-                raw_value = struct_parse(structs.Dwarf_dw_form[form], self.stream)
-                value = self._translate_attr_value(form, raw_value)
-            self.attributes[name] = AttributeValue(
-                name=name,
-                form=form,
-                value=value,
-                raw_value=raw_value,
-                offset=attr_offset,
-                indirection_length = indirection_length)
-
-        self.size = self.stream.tell() - self.offset
+        try:
+            structs = self.cu.structs
+            stream = self.stream
+
+            # A DIE begins with the abbreviation code. Read it and use it to
+            # obtain the abbrev declaration for this DIE.
+            # Note: here and elsewhere, preserve_stream_pos is used on operations
+            # that manipulate the stream by reading data from it.
+            self.abbrev_code = struct_parse(
+                structs.the_Dwarf_uleb128, stream, self.offset)
+
+            # This may be a null entry
+            if self.abbrev_code == 0:
+                self.size = stream.tell() - self.offset
+                return
+
+            abbrev_decl = self.cu.get_abbrev_table().get_abbrev(self.abbrev_code)
+            self.tag = abbrev_decl['tag']
+            self.has_children = abbrev_decl.has_children()
+
+            # Guided by the attributes listed in the abbreviation declaration, parse
+            # values from the stream.
+            for spec in abbrev_decl['attr_spec']:
+                form = spec.form
+                name = spec.name
+                attr_offset = stream.tell()
+                indirection_length = 0
+                # Special case here: the attribute value is stored in the attribute
+                # definition in the abbreviation spec, not in the DIE itself.
+                if form == 'DW_FORM_implicit_const':
+                    value = spec.value
+                    raw_value = value
+                # Another special case: the attribute value is a form code followed by the real value in that form
+                elif form == 'DW_FORM_indirect':
+                    (form, raw_value, indirection_length) = self._resolve_indirect()
+                    value = self._translate_attr_value(form, raw_value)
+                else:
+                    raw_value = structs.Dwarf_dw_form[form].parse_stream(stream)
+                    value = self._translate_attr_value(form, raw_value)
+                self.attributes[name] = AttributeValue(
+                    name=name,
+                    form=form,
+                    value=value,
+                    raw_value=raw_value,
+                    offset=attr_offset,
+                    indirection_length = indirection_length)
+
+            self.size = stream.tell() - self.offset
+        except ConstructError as e:
```

Comment:
What is this for? 

The extra indentation makes it difficult to review the actual change here; can it be done separately if needed?

Reply:
Before, the "read the attribute" value line was a call to `struct_parse`. That did two things:

 - seek if needed
 - set up a try/except block and translate `FieldError` into `ELFParseException`

I've replaced that line with a call to parse_stream() on the parser object, since there is no need to seek. Avoided a needless `if()` and an extra function call on the hot path. But the logic of translating FieldError to ELFParseException needed to be kept, so I've moved it to the DIE level. There was a measurable effect.

Reply:
Understood, thanks